test_report/src/report.rs

```diff
@@ -0,0 +1,268 @@
+use chrono::prelude::*;
+#[cfg(feature = "ruby")]
+use magnus::{value::ReprValue, Module, Object};
+use prost_wkt_types::Timestamp;
+use proto::test_context::test_run::{TestCaseRun, TestCaseRunStatus, TestResult, UploaderMetadata};
+use std::cell::RefCell;
+#[cfg(feature = "wasm")]
+use wasm_bindgen::prelude::wasm_bindgen;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct TestReport {
+    test_result: TestResult,
+}
+
+#[cfg_attr(feature = "wasm", wasm_bindgen)]
+#[cfg_attr(feature = "ruby", magnus::wrap(class = "TestExecution"))]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TestExecution {
+    id: Option<String>,
+    name: String,
+    classname: String,
+    file: String,
+    parent_name: String,
+    line: Option<i32>,
+    status: Status,
+    attempt_number: i32,
+    started_at: i64,
+    finished_at: i64,
+    output: String,
+}
+
+#[cfg_attr(feature = "wasm", wasm_bindgen)]
+#[cfg_attr(feature = "ruby", magnus::wrap(class = "Status"))]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Status {
+    Success,
+    Failure,
+    Skipped,
+}
+
+#[cfg(feature = "ruby")]
+impl Status {
+    fn new(status: String) -> Self {
+        match status.as_str() {
+            "success" => Status::Success,
+            "failure" => Status::Failure,
+            "skipped" => Status::Skipped,
+            _ => Status::Success,
+        }
+    }
+}
+
+impl Into<&str> for Status {
+    fn into(self) -> &'static str {
+        match self {
+            Status::Success => "success",
+            Status::Failure => "failure",
+            Status::Skipped => "skipped",
+        }
+    }
+}
+
+impl ToString for Status {
+    fn to_string(&self) -> String {
+        String::from(Into::<&str>::into(*self))
+    }
+}
+
+#[cfg(feature = "ruby")]
+impl Status {
+    pub fn to_string(&self) -> &str {
+        (*self).into()
+    }
+}
+
+#[cfg(feature = "ruby")]
+impl magnus::TryConvert for Status {
+    fn try_convert(val: magnus::Value) -> Result<Self, magnus::Error> {
+        let sval: String = val.funcall("to_s", ())?;
+        Ok(Status::new(sval))
+    }
+}
+
+#[cfg_attr(feature = "wasm", wasm_bindgen(getter_with_clone))]
+#[cfg_attr(feature = "ruby", magnus::wrap(class = "TestReport"))]
+#[derive(Debug, Clone, PartialEq)]
+pub struct MutTestReport(RefCell<TestReport>);
+
+#[cfg_attr(feature = "wasm", wasm_bindgen)]
+impl MutTestReport {
```

Comment:
Constructed a separate mutable test report as per the suggestions of `magnus`. It reduces the boilerplate needed because we do not need to have bindings for all of the stored types (the proto object in this case).